src/main/java/org/apache/ibatis/builder/annotation/AbstractProviderSqlSource.java

```diff
@@ -0,0 +1,99 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+package org.apache.ibatis.builder.annotation;
+
+import org.apache.ibatis.builder.BuilderException;
+import org.apache.ibatis.builder.SqlSourceBuilder;
+import org.apache.ibatis.mapping.BoundSql;
+import org.apache.ibatis.mapping.SqlSource;
+import org.apache.ibatis.reflection.ParamNameResolver;
+import org.apache.ibatis.session.Configuration;
+
+import java.lang.reflect.Method;
+
+/**
+ * @author Clinton Begin
+ * @author Kazuki Shimizu
+ * @author liuzh
+ * @since 3.4.6
+ */
+public abstract class AbstractProviderSqlSource implements SqlSource {
+
+  protected final Configuration configuration;
+  protected final SqlSourceBuilder sqlSourceParser;
+  protected final Class<?> providerType;
+  protected Method providerMethod;
+  protected String[] providerMethodArgumentNames;
+  protected Class<?>[] providerMethodParameterTypes;
+  protected ProviderContext providerContext;
+  protected Integer providerContextIndex;
+
+  /**
+   * @since 3.4.6
+   */
+  public AbstractProviderSqlSource(Configuration configuration, Object provider, Class<?> mapperType, Method mapperMethod) {
+    String providerMethodName;
+    try {
+      this.configuration = configuration;
+      this.sqlSourceParser = new SqlSourceBuilder(configuration);
+      this.providerType = (Class<?>) provider.getClass().getMethod("type").invoke(provider);
+      providerMethodName = (String) provider.getClass().getMethod("method").invoke(provider);
+
+      for (Method m : this.providerType.getMethods()) {
+        if (providerMethodName.equals(m.getName())) {
+          if (m.getReturnType() == String.class) {
+            if (providerMethod != null){
+              throw new BuilderException("Error creating SqlSource for SqlProvider. Method '"
+                      + providerMethodName + "' is found multiple in SqlProvider '" + this.providerType.getName()
+                      + "'. Sql provider method can not overload.");
+            }
+            this.providerMethod = m;
+            this.providerMethodArgumentNames = new ParamNameResolver(configuration, m).getNames();
+            this.providerMethodParameterTypes = m.getParameterTypes();
+          }
+        }
+      }
+    } catch (BuilderException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new BuilderException("Error creating SqlSource for SqlProvider.  Cause: " + e, e);
+    }
+    if (this.providerMethod == null) {
+      throw new BuilderException("Error creating SqlSource for SqlProvider. Method '"
+          + providerMethodName + "' not found in SqlProvider '" + this.providerType.getName() + "'.");
+    }
+    for (int i = 0; i< this.providerMethodParameterTypes.length; i++) {
```

Comment:
how about 

```java
if(this.providerContext != null) {
   Arrays.stream(this.providerMethodParameterTypes).filter(pt -> pt.equals(ProviderContext.class))
   .ifPresent(p -> throw new BuilderException(...));
}
```

also, while it may seem correct to do `someClass == otherClass` this is subtlety error prone. 2 classes are equal if they are of the same package and name *AND* use the same class loader. You might say: Well, of course these are in the same class loader! Application containers like `Tomcat` have a hierarchical class loader and this does not hold. Use `.equals`